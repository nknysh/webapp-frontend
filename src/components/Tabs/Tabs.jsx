import React, { useState, Fragment } from 'react';
import { map, propOr } from 'ramda';

import { propTypes, defaultProps } from './Tabs.props';
import { StyledTabs, Tab } from './Tabs.styles';

const renderLabel = label => <Tab key={label} label={label} />;

export const Tabs = ({ labels, children, ...props }) => {
  const [current, setCurrent] = useState(0);

  const content = propOr(children, current, children);

  const onChange = (e, value) => setCurrent(value);

  return (
    <Fragment>
      <StyledTabs classes={{ indicator: 'tab-line' }} value={current} className="tabs" onChange={onChange} {...props}>
        {map(renderLabel, labels)}
      </StyledTabs>
      {content}
    </Fragment>
  );
};

Tabs.propTypes = propTypes;
Tabs.defaultProps = defaultProps;

export default Tabs;
