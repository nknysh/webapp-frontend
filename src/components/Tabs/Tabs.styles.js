import styled, { css } from 'styled-components';
import { Tabs, Tab as BaseTab } from '@material-ui/core';

import theme from 'styles/theme';

export const StyledTabs = styled(Tabs)`
  background: ${theme.colors.whiteish};

  .tab-line {
    height: 1px;
    background: ${theme.primary};
  }
`;

export const Tab = styled(BaseTab)`
  color: ${theme.colors['gold-light']} !important;
  flex-grow: 1 !important;
  font-family: ${theme.defaultFont} !important;
  font-size: 12px !important;
  text-transform: uppercase;
  font-weight: ${theme.bold} !important
    ${({ selected }) =>
      selected &&
      css`
        color: ${theme.colors['gold-dark']} !important;
      `};
`;
