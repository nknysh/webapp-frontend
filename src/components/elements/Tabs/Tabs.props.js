import PropTypes from 'prop-types';

import { noop } from 'utils';

export const propTypes = {
  centered: PropTypes.bool,
  labels: PropTypes.array,
  current: PropTypes.number,
  onChange: PropTypes.func,
  onTabScrollButtonChange: PropTypes.func,
};

export const defaultProps = {
  labels: [],
  value: 0,
  onChange: noop,
  onTabScrollButtonChange: noop,
};
