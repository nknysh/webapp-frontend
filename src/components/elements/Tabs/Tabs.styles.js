import styled, { css } from 'styled-components';
import { Tabs, Tab as BaseTab } from '@material-ui/core';

import theme from 'styles/theme';

export const StyledTabs = styled(Tabs)`
  background: ${theme.backgrounds.secondary};
  border-bottom: 1px solid ${theme.primary};

  .tab-line {
    height: 1px;
    background: ${theme.primary};
  }
`;

export const Tab = styled(BaseTab)`
  color: ${theme.colors['gold-light']} !important;
  flex-grow: 1 !important;
  font-family: ${theme.fonts.defaultFont} !important;
  font-size: ${theme.fonts.sizes.default}px !important;
  text-transform: uppercase;
  font-weight: ${theme.fonts.bold} !important;
  ${({ selected }) =>
    selected &&
    css`
      color: ${theme.secondary} !important;
    `};
`;
