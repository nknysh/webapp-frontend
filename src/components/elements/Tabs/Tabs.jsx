import React, { useState, Fragment, useCallback, Children } from 'react';
import { propOr, isEmpty, partial, inc, gt, length, lt, dec, equals } from 'ramda';
import TabScrollButton from '@material-ui/core/Tabs/TabScrollButton';

import { mapWithIndex } from 'utils';

import { propTypes, defaultProps } from './Tabs.props';
import { StyledTabs, Tab } from './Tabs.styles';

const renderLabel = ({ tabClassname }, label, i) => <Tab key={i} className={tabClassname} label={label} />;

const renderTabArrow = ({ onTabScrollChange }) => {
  return props => {
    const { direction, onClick } = props;

    const onClickHandler = (...args) => {
      onTabScrollChange(direction);
      onClick(...args);
    };

    return <TabScrollButton {...props} onClick={onClickHandler} />;
  };
};

export const Tabs = ({ labels, children, onChange, value, tabClassname, onTabScrollButtonChange, ...props }) => {
  const [tab, setTab] = useState(value);

  const childrenArr = Children.toArray(children);
  const childrenLength = length(childrenArr);

  const content = propOr(childrenArr, tab, childrenArr);

  const onTabScrollChange = useCallback(
    direction => {
      onTabScrollButtonChange(direction);
      const next = inc(tab);
      const prev = dec(tab);

      const nextTab = equals('right', direction) ? (gt(next, childrenLength) ? tab : next) : lt(prev, 0) ? tab : prev;

      setTab(nextTab);
    },
    [tab, childrenArr, childrenLength]
  );

  const onTabSelect = useCallback(
    (e, value) => {
      setTab(value);
      onChange(value);
    },
    [onChange]
  );

  return (
    <Fragment>
      {!isEmpty(labels) && (
        <StyledTabs
          ScrollButtonComponent={renderTabArrow({ onTabScrollChange })}
          classes={{ indicator: 'tab-line' }}
          value={tab}
          className="tabs"
          onChange={onTabSelect}
          {...props}
        >
          {mapWithIndex(partial(renderLabel, [{ tabClassname }]), labels)}
        </StyledTabs>
      )}

      {content}
    </Fragment>
  );
};

Tabs.propTypes = propTypes;
Tabs.defaultProps = defaultProps;

export default Tabs;
