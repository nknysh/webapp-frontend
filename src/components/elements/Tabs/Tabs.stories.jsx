import React from 'react';
import { range } from 'ramda';

import { mapWithIndex } from 'utils';

import { storiesOf } from '@storybook/react';
import { withKnobs, number, boolean } from '@storybook/addon-knobs';

import Tabs from './Tabs';

storiesOf('Tabs', module)
  .addDecorator(withKnobs)
  .add('default', () => {
    const items = number('Number of crumbs', 2, {
      range: true,
      min: 1,
      max: 5,
      step: 1,
    });

    const renderLabel = number => `Crumb ${number}`;
    const renderContent = number => <h1>Content {number}</h1>;

    return (
      <Tabs centered={boolean('Centered', false)} labels={mapWithIndex(renderLabel, range(0, items))}>
        {mapWithIndex(renderContent, range(0, items))}
      </Tabs>
    );
  });
