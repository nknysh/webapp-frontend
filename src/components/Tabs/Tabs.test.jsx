import React from 'react';

import { Tabs } from './Tabs';

const getComponent = props => shallow(<Tabs labels={[]} {...props} />);

describe('<Tabs />', () => {
  describe('render', () => {
    it('matches snapshot', () => {
      expect(getComponent()).toMatchSnapshot();
      expect(getComponent({ labels: ['foo', 'bar'] })).toMatchSnapshot();
    });
  });
});
