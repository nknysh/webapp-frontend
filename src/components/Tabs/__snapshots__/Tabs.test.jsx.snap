// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> render matches snapshot 1`] = `<Fragment />`;

exports[`<Tabs /> render matches snapshot 2`] = `
<Fragment>
  <Styled(WithStyles(Tabs))
    className="tabs"
    classes={
      Object {
        "indicator": "tab-line",
      }
    }
    onChange={[Function]}
    value={0}
  >
    <Styled(WithStyles(Tab))
      key="foo"
      label="foo"
    />
    <Styled(WithStyles(Tab))
      key="bar"
      label="bar"
    />
  </Styled(WithStyles(Tabs))>
</Fragment>
`;
